[Reporting API] Hook up to Content-Security-Policy 'report-to' directive

brentfulgham opened the pull request:
#### 42f5a93823a7f087a800cd65c6bc0551dbeb55d3
<pre>
[Reporting API] Hook up to Content-Security-Policy &apos;report-to&apos; directive
<a href="https://bugs.webkit.org/show_bug.cgi?id=244143">https://bugs.webkit.org/show_bug.cgi?id=244143</a>
&lt;rdar://problem/98900892&gt;

Reviewed by Ryosuke Niwa.

Implement the stubs from Bug 243908 so that we can generate Reporting API reports for Content Security Policy rules.

This patch does the following:

1. Adds support for the &quot;Reporting-Endpoints&quot; header.
2. Exposes a Reporting object on Document and WorkerGlobalScope that can handle reports.
3. Implements a CSSViolationReportBody class (and JS binding)
4. Update the ContentSecurityPolicy implementation to generate a CSSViolationReportBody when the CSP directives include a report-to declaration.
5. Revises the CSP implementation to construct the JSON report body to match the new specification.

* Source/WTF/wtf/URL.cpp:
(WTF::URL::strippedForUseAsReferrerWithExplicitPort): Added.
* Source/WTF/wtf/URL.h:
* Source/WebCore/CMakeLists.txt: Update for new files.
* Source/WebCore/DerivedSources-input.xcfilelist: Ditto.
* Source/WebCore/DerivedSources-input.xcfilelist: Ditto.
* Source/WebCore/DerivedSources.make: Ditto.
* Source/WebCore/CMakeLists.txt: Ditto.
* Source/WebCore/Headers.cmake: Ditto.
* Source/WebCore/Modules/reporting/Report.cpp: Added.
* Source/WebCore/Modules/reporting/Report.h: Added.
* Source/WebCore/Modules/reporting/Report.idl: Added.
* Source/WebCore/Modules/reporting/ReportBody.cpp: Added.
* Source/WebCore/Modules/reporting/ReportBody.h: Added.
* Source/WebCore/Modules/reporting/ReportBody.idl: Added.
* Source/WebCore/Modules/reporting/ReportingClient.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.cpp: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.idl: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.idl: Added.
* Source/WebCore/Modules/reporting/ReportingScope.cpp: Added.
* Source/WebCore/Modules/reporting/ReportingScope.h: Added.
* Source/WebCore/Sources.txt: Update for new files.
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp: Added.
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h: Updated for new types.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::notifyReportObservers): Added.
(WebCore::Document::endpointURIForToken const): Added.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/SecurityPolicyViolationEvent.h: Move SecurityPolicyViolationEventDisposition
definition to new file to allow sharing with CSP.
* Source/WebCore/dom/SecurityPolicyViolationEvent.idl: Ditto.
* Source/WebCore/dom/SecurityPolicyViolationEventDisposition.h: Added.
* Source/WebCore/dom/SecurityPolicyViolationEventDisposition.idl: Added.
* Source/WebCore/dom/TaskSource.h: Added new Reporting Task type.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Revise to pass new arguments to CSP constructor.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument): Capture report-to endpoints for later use.
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::WorkerThreadableLoader::MainThreadBridge::MainThreadBridge): Register reporting observer.
* Source/WebCore/page/csp/CSPViolationReportBody.cpp: Added.
* Source/WebCore/page/csp/CSPViolationReportBody.h: Added.
* Source/WebCore/page/csp/CSPViolationReportBody.idl: Added.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::reportingClientForContext): Added.
(WebCore::ContentSecurityPolicy::ContentSecurityPolicy): Update to capture reporting client.
(WebCore::ContentSecurityPolicy::reportViolation const): Revise to create modern Reporting API
report, when used.
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp: Parse new report-to directive.
(WebCore::ContentSecurityPolicyDirectiveList::create): Check for report-uri and report-to when validating CSP.
(WebCore::ContentSecurityPolicyDirectiveList::parseReportTo): Added.
(WebCore::ContentSecurityPolicyDirectiveList::addDirective): Update to handle report-to.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveNames.cpp: Add report-to.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::isCSPDirectiveName): Update to handle report-to.
* Source/WebCore/platform/network/HTTPHeaderNames.in: Add &apos;Reporting-Endpoints&apos;
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::notifyReportObservers):
(WebCore::WorkerGlobalScope::endpointURIForToken const):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldInterruptLoadForCSPFrameAncestorsOrXFrameOptions): Parse Report-To Header.
(WebKit::NetworkResourceLoader::notifyReportObservers): Added to implement ReportClient interface.
(WebKit::NetworkResourceLoader::endpointURIForToken): Ditto.
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::contentSecurityPolicy): Initialize reporting client.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::doCrossOriginOpenerHandlingOfResponse): Initialize reporting client.
(WebKit::NetworkResourceLoader::endpointURIForToken const): Added to implement ReportClient interface.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp: Updated for new CSPViolationReportBody and Report types.
* Source/WebKit/Shared/WebCoreArgumentCoders.h
* Source/WebKit/WebProcess/WebPage.cpp:
(WebKit::WebPage::notifyReportObservers): Added to implement ReportClient interface.
* Source/WebKit/WebProcess/WebPage.h:
* Source/WebKit/WebProcess/WebPage.messages.in: Added new &apos;NotifyReportObservers&apos; IPC message.
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(webView:resource:willSendRequest:redirectResponse:fromDataSource:): Filter unique report ID&apos;s in logging
to avoid test failures.
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTR::InjectedBundlePage::willSendRequestForFrame):  Filter unique report ID&apos;s in logging to avoid test failures.

Canonical link: <a href="https://commits.webkit.org/253966@main">https://commits.webkit.org/253966@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5049057a2b39f449a87addc700e439d9e43987b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87646 "failed 1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31733 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/18403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96852 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/150614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30091 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79745 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93264 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74405 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24138 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/18403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27790 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/18403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27753 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/18403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2790 "Committed 253966@main") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29446 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/37200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29352 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/18403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on LayoutTests/TestExpectations:
```diff
@@ -950,7 +951,7 @@ imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/l
 imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/cross-origin-to-whom-part-2.window.html [ Failure Pass ]
 imported/w3c/web-platform-tests/html/semantics/forms/the-selectmenu-element/selectmenu-popup-position-with-zoom.tentative.html [ Failure Pass ]
-imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-ancestors.https.sub.html [ Failure Pass ]
+#imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-ancestors.https.sub.html [ Failure Pass ]
```
Whoops -- didn't meant to include this.

brentfulgham commented on Source/WebCore/dom/Document.h:
```diff
@@ -348,18 +349,18 @@ class Document
     , public Supplementable<Document>
     , public Logger::Observer
-    , public CanvasObserver {
+, public CanvasObserver {
```
Whoops -- fixing.

brentfulgham reviewed (commented):
I need to fix some whitespace changes that got in by mistake.

rniwa commented on Source/WebCore/Modules/reporting/ReportBody.h:
```diff
@@ -25,23 +25,40 @@
+    // More to come
+};
+
+class WEBCORE_EXPORT ReportBody : public RefCounted<ReportBody> {
```
Don't we need encode/decode for ReportBody as well?

brentfulgham replied:
We don't ever encode a `ReportBody` (which is kind of an abstract class). Right now, we only encode a `CSPViolationReportBody`, and in the future we will encode a couple of others.

rniwa commented on Source/WebCore/Modules/reporting/Report.h:
```diff
@@ -31,23 +31,55 @@
-    String m_type;
+    AtomString m_type;
     String m_url;
     RefPtr<ReportBody> m_body;
```
Can ReportBody be empty??? If not, why don't we use Ref instead?

brentfulgham replied:
According to the spec, The `ReportBody` in a `Report` is nullable, so I think it's appropriate to be a `RefPtr`, though I don't believe we ever construct a `Report` without a `ReportBody`.

rniwa commented on Source/WebCore/page/csp/CSPViolationReportBody.cpp:
```diff
@@ -0,0 +1,186 @@
+    return m_columnNumber;
+}
+
+Ref<FormData> CSPViolationReportBody::createReportFormDataForViolation(const CSPInfo& info, bool reportToFormat, bool isReportOnly,
```
reportToFormat sounds like "reporting to format".
How about usesReportTo as in the caller?

brentfulgham replied:
Will do!

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+        if (m_reportingClient) {
+            m_reportingClient->notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+
+            for (auto token : violatedDirectiveList.reportToTokens()) {
```
auto&?

brentfulgham replied:
Will do!

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+            auto& reporting = reportingForContext(*m_scriptExecutionContext);
+            reporting.notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+            
+            for (auto token : violatedDirectiveList.reportToTokens()) {
```
auto&?

brentfulgham replied:
Done!

rniwa commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -3220,4 +3223,55 @@ void ArgumentCoder<PixelBuffer>::encode<Encoder>(Encoder&, const PixelBuffer&);
+        return CSPViolationReportBody::decode(decoder);
+    }
+
+    RELEASE_ASSERT_NOT_REACHED();
```
I don't think we want to release assert here.
If a rogue web content process sends a bad IPC, we shouldn't crash the entire Safari.

brentfulgham replied:
Good point -- I'll treat it as a serialization failure and return std::nullopt and use a debug assert.

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4600,6 +4602,18 @@ void WebPage::enqueueSecurityPolicyViolationEvent(FrameIdentifier frameID, Secur
+    auto* coreFrame = frame->coreFrame();
+    if (!coreFrame)
+        return;
+    if (auto* document = coreFrame->document())
```
Please use RefPtr.

brentfulgham replied:
Done.

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/content-security-policy/gen/top.http-rp/worker-src-self/sharedworker-import.http-expected.txt:
```diff
@@ -4,11 +4,11 @@ PASS Content Security Policy: Expects allowed for sharedworker-import to same-ht
 PASS Content Security Policy: Expects allowed for sharedworker-import to same-http origin and no-redirect redirection from http context.: securitypolicyviolation
 PASS Content Security Policy: Expects blocked for sharedworker-import to cross-http origin and keep-origin redirection from http context.
-PASS Content Security Policy: Expects blocked for sharedworker-import to cross-http origin and keep-origin redirection from http context.: securitypolicyviolation
+FAIL Content Security Policy: Expects blocked for sharedworker-import to cross-http origin and keep-origin redirection from http context.: securitypolicyviolation assert_equals: One violation event should be fired expected 1 but got 0
```
This seems like a regression??

brentfulgham replied:
I'm uploading a revised patch for EWS, but will investigate this and resolve in a follow-up.

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/content-security-policy/gen/top.http-rp/worker-src-self/sharedworker-import.https-expected.txt:
```diff
@@ -4,11 +4,11 @@ PASS Content Security Policy: Expects allowed for sharedworker-import to same-ht
 PASS Content Security Policy: Expects allowed for sharedworker-import to same-https origin and no-redirect redirection from https context.: securitypolicyviolation
 PASS Content Security Policy: Expects blocked for sharedworker-import to cross-https origin and keep-origin redirection from https context.
-PASS Content Security Policy: Expects blocked for sharedworker-import to cross-https origin and keep-origin redirection from https context.: securitypolicyviolation
+FAIL Content Security Policy: Expects blocked for sharedworker-import to cross-https origin and keep-origin redirection from https context.: securitypolicyviolation assert_equals: One violation event should be fired expected 1 but got 0
```
Ditto.

brentfulgham replied:
I'm uploading a revised patch for EWS, but will investigate this and resolve in a follow-up.

rniwa commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -843,6 +843,28 @@ String URL::strippedForUseAsReferrer() const
     );
 }
 
+String URL::strippedForUseAsReferrerWithExplicitPort() const
```
Why are we adding this new function? Please explain in the commit message.

brentfulgham replied:
Will do. In short, WPT expects the port name be included in the URL supplied to the report. Perhaps a better name would be "URL::stringSuitableForReporting" or "URL::stringIncludingExplicitPort".

rniwa replied:
Huh, that's so bizarre. Is that mentioned anywhere in the spec?

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -26,24 +26,48 @@
+
+static Ref<Reporting> reportingForContext(ScriptExecutionContext& scriptExecutionContext)
+{
+    if (is<Document>(scriptExecutionContext))
```
Why not dynamicDowncast?

brentfulgham replied:
Good point -- I'll revise to do so.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -26,24 +26,48 @@
+    if (is<Document>(scriptExecutionContext))
+        return downcast<Document>(scriptExecutionContext).reporting();
+
+    if (is<WorkerGlobalScope>(scriptExecutionContext))
```
Ditto.

brentfulgham replied:
Done.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -53,15 +77,52 @@ ReportingObserver::~ReportingObserver()
 
 void ReportingObserver::observe()
 {
+    m_reporting->registerReportingObserver(*this);
```
Who keeps the observer alive?
Otherwise, we'd just GC this observer at any moment.

brentfulgham replied:
The reporting observer is created in JS by the page author. I assumed the lifecycle would be controlled by the script or page lifetime, but perhaps this is not right? Maybe it should be an ActiveDOMObject?

rniwa replied:
That's not quite right. So long as observer is observing anything at all, it's possible for the observer to be called back in the future. In effect, it needs to be kept alive so long as the global object is alive. As we discussed below, using `GenerateIsReachable=ImplScriptExecutionContext` will do the trick.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.idl:
```diff
@@ -28,6 +28,7 @@
 typedef sequence<Report> ReportList;
 
 [
+    // ActiveDOMObject,
```
I don't think this can be an active DOM object.
We're perpetually waiting for a new CSP violation, for example.
In effect, we need to "leak" any reporting observer whenever observe is called.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.idl:
```diff
@@ -38,6 +39,6 @@ typedef sequence<Report> ReportList;
 
 dictionary ReportingObserverOptions {
-    sequence<DOMString>? types;
+    sequence<[AtomString] DOMString>? types;
```
Sigh... this should really have been an argument of observe, not the constructor.

brentfulgham replied:
:-(

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.h:
```diff
@@ -25,36 +25,42 @@
 class ScriptExecutionContext;
 
-class ReportingObserver final : public RefCounted<ReportingObserver> {
+class ReportingObserver final : public RefCounted<ReportingObserver>, public ContextDestructionObserver, public CanMakeWeakPtr<ReportingObserver>  {
```
Do we really need a WeakPtr? I feel like CheckedPtr / CheckedRef is sufficient here.

brentfulgham replied:
That's a good point. If I stored `Ref`'s to `ReportingObservers` in the `Reporting` object (owned by the `Document` or `WorkerOrWorkletGlobalScope`) it would prevent anything form being GC'd until they had been explicitly unregistered or the document was closed. I'll try that approach.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -53,15 +77,52 @@ ReportingObserver::~ReportingObserver()
+void ReportingObserver::appendQueuedReportIfCorrectType(const Ref<Report>& report)
 {
-    return { };
+    // Step 4.3.1
```
Step 4.3.1. of what??

brentfulgham replied:
I added relevant URLs.

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -53,15 +77,52 @@ ReportingObserver::~ReportingObserver()
 }
 
-Vector<Report> ReportingObserver::takeRecords()
+void ReportingObserver::appendQueuedReportIfCorrectType(const Ref<Report>& report)
```
Is this https://w3c.github.io/reporting/#add-report?
If so, dropping a URL would be helpful.

brentfulgham replied:
Done

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -53,15 +77,52 @@ ReportingObserver::~ReportingObserver()
+    
+    auto reportsForCallback = takeRecords();
+    // Step 4.3.4: Queue a task to § 4.4
+    // Step 4.4: Invoke reporting observers with notify list with a copy of global’s registered reporting observer list.
```
Use context->eventLoop().queueTask(~)

brentfulgham replied:
Done.

rniwa commented on Source/WebCore/Modules/reporting/Reporting.cpp:
```diff
@@ -0,0 +1,137 @@
+    m_queuedReports.clear();
+}
+
+void Reporting::notifyReportObservers(Ref<Report>&& report)
```
Add a URL to https://w3c.github.io/reporting/#notify-observers?

brentfulgham replied:
Done.

rniwa commented on Source/WebCore/Modules/reporting/Reporting.h:
```diff
@@ -0,0 +1,69 @@
+class ReportingObserver;
+class ScriptExecutionContext;
+
+class WEBCORE_EXPORT Reporting final : public RefCounted<Reporting>, public ContextDestructionObserver {
```
Why don't we call this ReportingScope to match the terminology in the spec:
https://w3c.github.io/reporting/#notify-observers

brentfulgham replied:
Good idea!

rniwa commented on Source/WebCore/Modules/reporting/Reporting.h:
```diff
@@ -0,0 +1,69 @@
+class ScriptExecutionContext;
+
+class WEBCORE_EXPORT Reporting final : public RefCounted<Reporting>, public ContextDestructionObserver {
+    WTF_MAKE_FAST_ALLOCATED;
```
Maybe iso-heap this?

brentfulgham replied:
Yes -- I overlooked this.

rniwa commented on Source/WebCore/Modules/reporting/Reporting.h:
```diff
@@ -0,0 +1,69 @@
+private:
+    explicit Reporting(ScriptExecutionContext&);
+
+    Vector<WeakPtr<ReportingObserver>> m_reportingObservers;
```
We probably need to visit every reporting observer whenever the global object's visit children is called.
Alternatively, each ReportingObserver should be considered to be reachable from DOMWindow / WorkerGlobalScope.

brentfulgham replied:
Hmm. Is that done by adding something to the IDL, like `GenerateIsReachable` or `GenerateAddOpaqueRoot` or something?

brentfulgham replied:
Maybe: `GenerateIsReachable=ImplScriptExecutionContext`?

rniwa replied:
Yeah, `GenerateIsReachable=ImplScriptExecutionContext` will do the trick.

brentfulgham replied:
I apologize -- you said this in Slack and I missed it :-)

rniwa commented on Source/WebCore/Modules/reporting/Reporting.cpp:
```diff
@@ -0,0 +1,137 @@
+
+void Reporting::appendQueuedReportForRelevantType(ReportingObserver& observer)
+{
+    for (auto report : m_queuedReports)
```
auto&?

brentfulgham replied:
Will do!

rniwa commented on Source/WebCore/Modules/reporting/Reporting.cpp:
```diff
@@ -0,0 +1,137 @@
+        if (!tokenizer.consume('='))
+            break;
+
+        String urlPath = tokenizer.consumeTokenOrQuotedString().stripLeadingAndTrailingCharacters(isQuote);
```
It seems that we're supposed to parse this URL and continue / bail if it doesn't parse (step 5.2)
https://w3c.github.io/reporting/#process-header

brentfulgham replied:
Ugh! I missed that. Fixing.

rniwa commented on Source/WebCore/Modules/reporting/Reporting.cpp:
```diff
@@ -0,0 +1,137 @@
+
+        String urlPath = tokenizer.consumeTokenOrQuotedString().stripLeadingAndTrailingCharacters(isQuote);
+        
+        reportingEndpoints.set(WTFMove(reportTo), WTFMove(urlPath));
```
It seems that we're also supposed to skip any URL that's not "potentially trustworthy" (step 5.3):
https://w3c.github.io/reporting/#process-header

rniwa commented on Source/WebCore/Modules/reporting/ReportingObserver.cpp:
```diff
@@ -53,15 +80,63 @@ ReportingObserver::~ReportingObserver()
+    // Step 4.3.3:
+    m_queuedReports.append(report);
+    
+    auto reportsForCallback = takeRecords();
```
This is not right. Taking the reports must happen inside the queued task, not beforehand like this.

brentfulgham replied:
I see. Fixed.

rniwa commented on Source/WebCore/Modules/reporting/ReportingScope.cpp:
```diff
@@ -0,0 +1,159 @@
+    // Step 4.2.3-4: If scope’s report buffer now contains more than 100 reports with type equal to type, remove the earliest item with type equal to type in the report buffer.
+    // FIXME(244369): When we support more than one time of report, remove only the specific type of report.
+    if (m_queuedReports.size() > 100)
+        m_queuedReports.remove(0);
```
This is rather inefficient with Vector. Why not use Deque instead?

brentfulgham replied:
Good point -- I should have done that from the start.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+
+    Vector<String> reportURIs;
+    if (usesReportTo) {
+        if (m_reportingClient) {
```
When is this member variable null?

brentfulgham replied:
CSP running in the WebContent process has a ScriptExecutionContext. CSP running in the Network Process does not, but has a registered ReportingClient that it uses to relay state to the relevant WCP.

One possibility I considered was to have everything implement the ReportingClient interface (Document, Worker, and Clients), and do the right thing depending on context (CSP would register the Document or Worker in WCP, and the NetworkResourceLoader in the NetworkProcess.

That might clean up some of the duplicated code a bit.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+            m_reportingClient->notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+
+            for (auto& token : violatedDirectiveList.reportToTokens()) {
+                auto reportToURI = m_reportingClient->endpointForToken(token);
```
Maybe we should call this function endpointURIForToken?

brentfulgham replied:
Seems reasonable. Fixed.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -750,6 +756,17 @@ void ContentSecurityPolicy::reportViolation(const ContentSecurityPolicyDirective
 
+static ReportingScope& reportingScopeForContext(ScriptExecutionContext& scriptExecutionContext)
+{
+    if (auto document = dynamicDowncast<Document>(scriptExecutionContext))
```
This should be auto* or RefPtr.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -750,6 +756,17 @@ void ContentSecurityPolicy::reportViolation(const ContentSecurityPolicyDirective
+    if (auto document = dynamicDowncast<Document>(scriptExecutionContext))
+        return document->reportingScope();
+
+    if (auto workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(scriptExecutionContext))
```
Ditto.

brentfulgham replied:
Fixed.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+                reportURIs.append(WTFMove(reportToURI));
+            }
+        } else {
+            auto& reporting = reportingScopeForContext(*m_scriptExecutionContext);
```
Please use Ref or RefPtr.

brentfulgham replied:
Fixed.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
+            auto& reporting = reportingScopeForContext(*m_scriptExecutionContext);
+            reporting.notifyReportObservers(Report::create(CSPViolationReportBody::cspReportType(), info.documentURI, CSPViolationReportBody::create(SecurityPolicyViolationEventInit { violationEventInit })));
+            
+            for (auto& token : violatedDirectiveList.reportToTokens()) {
```
It's a bit sad that we're duplicating almost identical code here.
Maybe we can use template function to share code?
Alternatively, maybe ReportingScope should implement ReportingClient interface.

brentfulgham replied:
I think I've cleaned this up now.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.h:
```diff
@@ -240,6 +243,7 @@ class ContentSecurityPolicy {
     // We can never have both a script execution context and a ContentSecurityPolicyClient.
     ScriptExecutionContext* m_scriptExecutionContext { nullptr };
     ContentSecurityPolicyClient* m_client { nullptr };
+    ReportingClient* m_reportingClient { nullptr };
```
Can we use CheckedPtr or WeakPtr instead?

brentfulgham replied:
Yes! Done.

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -798,6 +817,10 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
     // FIXME: Is it policy to not use the status code for HTTPS, or is that a bug?
     unsigned short httpStatusCode = m_selfSourceProtocol == "http"_s ? m_httpStatusCode : 0;
+    
+    // WPT says modern tests expect valid status code, regardless of protocol:
```
Which WPT?

rniwa commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -808,13 +831,31 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
     violationEventInit.effectiveDirective = effectiveViolatedDirective;
     violationEventInit.originalPolicy = violatedDirectiveList.header();
     violationEventInit.sourceFile = info.sourceFile;
+    // WPT expects this behavior:
```
Ditto

rniwa reviewed (approved):
r=me

youennf committed:
[Reporting API] Hook up to Content-Security-Policy 'report-to' directive
https://bugs.webkit.org/show_bug.cgi?id=244143
<rdar://problem/98900892>

Reviewed by Ryosuke Niwa.

Implement the stubs from Bug 243908 so that we can generate Reporting API reports for Content Security Policy rules.

This patch does the following:

1. Adds support for the "Reporting-Endpoints" header.
2. Exposes a Reporting object on Document and WorkerGlobalScope that can handle reports.
3. Implements a CSSViolationReportBody class (and JS binding)
4. Update the ContentSecurityPolicy implementation to generate a CSSViolationReportBody when the CSP directives include a report-to declaration.
5. Revises the CSP implementation to construct the JSON report body to match the new specification.

* Source/WTF/wtf/URL.cpp:
(WTF::URL::strippedForUseAsReferrerWithExplicitPort): Added.
* Source/WTF/wtf/URL.h:
* Source/WebCore/CMakeLists.txt: Update for new files.
* Source/WebCore/DerivedSources-input.xcfilelist: Ditto.
* Source/WebCore/DerivedSources-input.xcfilelist: Ditto.
* Source/WebCore/DerivedSources.make: Ditto.
* Source/WebCore/CMakeLists.txt: Ditto.
* Source/WebCore/Headers.cmake: Ditto.
* Source/WebCore/Modules/reporting/Report.cpp: Added.
* Source/WebCore/Modules/reporting/Report.h: Added.
* Source/WebCore/Modules/reporting/Report.idl: Added.
* Source/WebCore/Modules/reporting/ReportBody.cpp: Added.
* Source/WebCore/Modules/reporting/ReportBody.h: Added.
* Source/WebCore/Modules/reporting/ReportBody.idl: Added.
* Source/WebCore/Modules/reporting/ReportingClient.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.cpp: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.idl: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.idl: Added.
* Source/WebCore/Modules/reporting/ReportingScope.cpp: Added.
* Source/WebCore/Modules/reporting/ReportingScope.h: Added.
* Source/WebCore/Sources.txt: Update for new files.
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp: Added.
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h: Updated for new types.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::notifyReportObservers): Added.
(WebCore::Document::endpointURIForToken const): Added.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/SecurityPolicyViolationEvent.h: Move SecurityPolicyViolationEventDisposition
definition to new file to allow sharing with CSP.
* Source/WebCore/dom/SecurityPolicyViolationEvent.idl: Ditto.
* Source/WebCore/dom/SecurityPolicyViolationEventDisposition.h: Added.
* Source/WebCore/dom/SecurityPolicyViolationEventDisposition.idl: Added.
* Source/WebCore/dom/TaskSource.h: Added new Reporting Task type.
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Revise to pass new arguments to CSP constructor.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument): Capture report-to endpoints for later use.
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::WorkerThreadableLoader::MainThreadBridge::MainThreadBridge): Register reporting observer.
* Source/WebCore/page/csp/CSPViolationReportBody.cpp: Added.
* Source/WebCore/page/csp/CSPViolationReportBody.h: Added.
* Source/WebCore/page/csp/CSPViolationReportBody.idl: Added.
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::reportingClientForContext): Added.
(WebCore::ContentSecurityPolicy::ContentSecurityPolicy): Update to capture reporting client.
(WebCore::ContentSecurityPolicy::reportViolation const): Revise to create modern Reporting API
report, when used.
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp: Parse new report-to directive.
(WebCore::ContentSecurityPolicyDirectiveList::create): Check for report-uri and report-to when validating CSP.
(WebCore::ContentSecurityPolicyDirectiveList::parseReportTo): Added.
(WebCore::ContentSecurityPolicyDirectiveList::addDirective): Update to handle report-to.
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveNames.cpp: Add report-to.
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::isCSPDirectiveName): Update to handle report-to.
* Source/WebCore/platform/network/HTTPHeaderNames.in: Add 'Reporting-Endpoints'
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::notifyReportObservers):
(WebCore::WorkerGlobalScope::endpointURIForToken const):
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldInterruptLoadForCSPFrameAncestorsOrXFrameOptions): Parse Report-To Header.
(WebKit::NetworkResourceLoader::notifyReportObservers): Added to implement ReportClient interface.
(WebKit::NetworkResourceLoader::endpointURIForToken): Ditto.
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::contentSecurityPolicy): Initialize reporting client.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::doCrossOriginOpenerHandlingOfResponse): Initialize reporting client.
(WebKit::NetworkResourceLoader::endpointURIForToken const): Added to implement ReportClient interface.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp: Updated for new CSPViolationReportBody and Report types.
* Source/WebKit/Shared/WebCoreArgumentCoders.h
* Source/WebKit/WebProcess/WebPage.cpp:
(WebKit::WebPage::notifyReportObservers): Added to implement ReportClient interface.
* Source/WebKit/WebProcess/WebPage.h:
* Source/WebKit/WebProcess/WebPage.messages.in: Added new 'NotifyReportObservers' IPC message.
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(webView:resource:willSendRequest:redirectResponse:fromDataSource:): Filter unique report ID's in logging
to avoid test failures.
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp:
(WTR::InjectedBundlePage::willSendRequestForFrame):  Filter unique report ID's in logging to avoid test failures.

Canonical link: https://commits.webkit.org/253966@main
